File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/barrage/view/EmojiSpan.java
package com.trtc.uikit.livekit.component.barrage.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.text.style.ImageSpan;

public class EmojiSpan extends ImageSpan {
    private int mEmojiTranslateY;

    public EmojiSpan(Drawable drawable, int translateY) {
        super(drawable);
        this.mEmojiTranslateY = translateY;
    }

    public EmojiSpan(Context context, int resourceId) {
        super(context, resourceId);
    }

    @Override
    public int getSize(Paint paint, CharSequence text, int start, int end, Paint.FontMetricsInt fontMetricsInt) {
        Drawable drawable = this.getDrawable();
        Rect rect = drawable.getBounds();
        if (fontMetricsInt != null) {
            Paint.FontMetricsInt fmPaint = paint.getFontMetricsInt();
            int fontHeight = fmPaint.bottom - fmPaint.top;
            int drHeight = rect.bottom - rect.top;
            int top = drHeight / 2 - fontHeight / 4;
            int bottom = drHeight / 2 + fontHeight / 4;
            fontMetricsInt.ascent = -bottom;
            fontMetricsInt.top = -bottom;
            fontMetricsInt.bottom = top;
            fontMetricsInt.descent = top;
        }
        return rect.right;
    }

    @Override
    public void draw(Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom,
                     Paint paint) {
        Drawable drawable = this.getDrawable();
        canvas.save();
        int transY = (bottom - top - drawable.getBounds().bottom) / 2 + top - mEmojiTranslateY;
        canvas.translate(x, (float) transY);
        drawable.draw(canvas);
        canvas.restore();
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/barrage/view/EmojiSpan.kt
package com.trtc.uikit.livekit.component.barrage.view

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.drawable.Drawable
import android.text.style.ImageSpan
import androidx.core.graphics.withSave

class EmojiSpan @JvmOverloads constructor(
    drawable: Drawable,
    private val emojiTranslateY: Int = 0
) : ImageSpan(drawable) {

    override fun getSize(
        paint: Paint,
        text: CharSequence,
        start: Int,
        end: Int,
        fm: Paint.FontMetricsInt?
    ) = drawable.bounds.right.also {
        fm?.run {
            val fontHeight = paint.fontMetricsInt.run { bottom - top }
            val drHeight = drawable.bounds.bottom
            val centerY = drHeight / 2 - fontHeight / 4
            ascent = -centerY
            top = -centerY
            bottom = centerY
            descent = centerY
        }
    }

    override fun draw(
        canvas: Canvas,
        text: CharSequence,
        start: Int,
        end: Int,
        x: Float,
        top: Int,
        y: Int,
        bottom: Int,
        paint: Paint
    ) {
        canvas.run {
            withSave {
                translate(x, ((bottom - top - drawable.bounds.bottom) / 2 + top - emojiTranslateY).toFloat())
                drawable.draw(this)
            }
        }
    }
}
